### PR TITLE
Téléphone obligatoire dans formulaire de vérification

### PR DESCRIPTION
**Pull request by yaf**
Rendre obligatoire le numéro de téléphone d'un usager dans le cas d'un rendez-vous téléphonique (#1113)

Cas pris en charge
- [x] Tunnel RDV coté usager, uniquement dans le formulaire
- [x] Tunnel RDV coté usager, coté serveur
- [x] Tunnel RDV coté agent

J'ai ajouté une notion de `error_path` car il y avait un souci dans l'enchainement « vérification du téléphone obligatoire, puis modification de l'usager depuis le tunnel de RDV coté agent. Au moment du retour, il y avait un souci d'url.

Pour le reste, je me suis inscrit dans les Form Models. Pas très confortable de voir l'héritage, ceci dit, ça reste lisible. C'est sans doute un point d'amélioration à explorer plus tard.

### CODE REVIEW

**Review by adipasquale — CHANGES_REQUESTED** [review]
c'est tres cool comme PR, content que les form models aient pu supporter cette évolution sans trop de modif.

1. Je mets en bloquant car ça la prise de RDV téléphonique côté usager m'a l'air cassée : même si tu mets un numéro de téléphone tu ne peux pas avancer. Il me semble que c'est parce que tu fais `@user_attributes[:phone_number]` avec un symbole plutot qu'une string. une solution peut etre de modifier la ligne en dessous `@user_attributes = @attributes[:user].with_indifferent_access`. sacrés symboles ruby

2. pas bloquant : un autre souci qu'il serait sympa de corriger sur le tunnel rdv coté usager : si tu forces la soumission du formulaire sans numero de telephone en ignorant l'attribut HTML required (par ex en le retirant avec l'inspecteur), tu ne vois pas de message d'erreur.  il me semble qu'il manque un affichage des erreurs dans le partial `users/user/form` ou bien dans la vue step.  C'est peu probable que required ne soit pas respecté mais je pense que ca serait cool de corriger

3. pas bloquant, suggestion : coté agent, bien vu pour le pb lors du refresh apres update d'un user dans la modale. Le souci est en effet qu'apres soumission la modale fait un refresh sur le `request.referer`, qui a ce moment vaut la valeur apres soumission du formulaire POST, donc une page irrechargeable. Ce pb est assez embetant dans le flow des tunnels car c'est quand meme possible que les agents aient envie de recharger. ta solution fonctionne bien pour cette step, dans ce cas. je me suis demandé si on ne pourrait pas faire une solution un peu plus générique : en ne soumettant plus les formulaires en POST mais en GET.  Je te propose un stash ici que tu peux tester et qui a l'air de fonctionner, mais il cassera peut etre un ou deux tests et j'ai peut etre raté qques edge cases. On peut tout a fait laisser comme tu as fait et ou decaler ca dans une autre PR, c'est juste une piste :

```
diff --git a/app/controllers/admin/rdv_wizard_steps_controller.rb b/app/controllers/admin/rdv_wizard_steps_controller.rb
index 5cc5107c..6f7287ee 100644
--- a/app/controllers/admin/rdv_wizard_steps_controller.rb
+++ b/app/controllers/admin/rdv_wizard_steps_controller.rb
@@ -22,8 +22,6 @@ class Admin::RdvWizardStepsController < AgentAuthController
     authorize(@rdv_wizard.rdv, :create?)
     if @rdv_wizard.save
       redirect_to @rdv_wizard.success_path, @rdv_wizard.success_flash
-    elsif @rdv_wizard.error_path.present?
-      redirect_to @rdv_wizard.error_path, flash: { error: @rdv_wizard.errors.full_messages.join(", ") }
     else
       render current_step
     end
diff --git a/app/form_models/admin/rdv_wizard_form.rb b/app/form_models/admin/rdv_wizard_form.rb
index 960dd82b..fa8c7886 100644
--- a/app/form_models/admin/rdv_wizard_form.rb
+++ b/app/form_models/admin/rdv_wizard_form.rb
@@ -69,10 +69,6 @@ module Admin::RdvWizardForm
     def success_path
       new_admin_organisation_rdv_wizard_step_path(@organisation, step: 3, **to_query)
     end
-
-    def error_path
-      new_admin_organisation_rdv_wizard_step_path(@organisation, step: 2, **to_query)
-    end
   end
 
   class Step3 < Step2
diff --git a/app/views/admin/rdv_wizard_steps/step1.html.slim b/app/views/admin/rdv_wizard_steps/step1.html.slim
index 82e8c4e4..6286d0fb 100644
--- a/app/views/admin/rdv_wizard_steps/step1.html.slim
+++ b/app/views/admin/rdv_wizard_steps/step1.html.slim
@@ -14,7 +14,8 @@
               = "Commence le : #{l(@rdv.starts_at, format: :human)}"
       .card-body
         = render partial: 'layouts/model_errors', locals: { model: @rdv_wizard }
-        = simple_form_for(@rdv_wizard, as: :rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation, step: 1)) do |f|
+        = simple_form_for(@rdv_wizard, as: :rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation), method: :get) do |f|
+          = hidden_field_tag "step", "1"
           = f.input_field :starts_at, as: :hidden
           = f.input_field :lieu_id, as: :hidden
           = f.input_field :context, as: :hidden
diff --git a/app/views/admin/rdv_wizard_steps/step2.html.slim b/app/views/admin/rdv_wizard_steps/step2.html.slim
index 5fb7b91a..5d3af81d 100644
--- a/app/views/admin/rdv_wizard_steps/step2.html.slim
+++ b/app/views/admin/rdv_wizard_steps/step2.html.slim
@@ -23,7 +23,8 @@
           label Usagers
           .js-users-spinner.spinner-border.d-none &nbsp;
 
-        = simple_form_for(@rdv_wizard, as: :rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation, step: 2)) do |f|
+        = simple_form_for(@rdv_wizard, as: :rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation), method: :get) do |f|
+          = hidden_field_tag "step", "2"
           - @rdv.users.each do |user|
             / https://stackoverflow.com/questions/2327621/creating-a-has-many-association-in-a-hidden-field
             = f.hidden_field "user_ids][", value: user.id
diff --git a/app/views/admin/rdv_wizard_steps/step3.html.slim b/app/views/admin/rdv_wizard_steps/step3.html.slim
index 6212a288..b3000167 100644
--- a/app/views/admin/rdv_wizard_steps/step3.html.slim
+++ b/app/views/admin/rdv_wizard_steps/step3.html.slim
@@ -17,8 +17,9 @@
           = users_inline_list_for_agents(@rdv.users)
 
       .card-body
-        = simple_form_for(@rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation, step: 3)) do |f|
+        = simple_form_for(@rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation), method: :get) do |f|
           = render(partial: 'layouts/model_errors', locals: { model: @rdv_wizard, f: f })
+          = hidden_field_tag "step", "3"
           = collapsible_form_fields_for_warnings(@rdv_wizard) do
             = f.association :motif, as: :hidden,  wrapper_html: { class: 'm-0' }
             = f.input :context, as: :hidden, wrapper_html: { class: 'm-0' }
diff --git a/config/routes.rb b/config/routes.rb
index d65e048e..a8888ca7 100644
--- a/config/routes.rb
+++ b/config/routes.rb
@@ -143,7 +143,9 @@ Rails.application.routes.draw do
           post :reinvite, on: :member
         end
         resource :merge_users, only: [:new, :create]
-        resource :rdv_wizard_step, only: [:new, :create]
+        resource :rdv_wizard_step, only: [:new] do
+          get :create
+        end
         devise_for :agents, controllers: { invitations: "admin/invitations_devise" }, only: :invitations
       end
 

```

**Comment by yaf** [comment]
> Je mets en bloquant car ça la prise de RDV téléphonique côté usager m'a l'air cassée : même si tu mets un numéro de téléphone tu ne peux pas avancer. Il me semble que c'est parce que tu fais @user_attributes[:phone_number] avec un symbole plutot qu'une string. une solution peut etre de modifier la ligne en dessous @user_attributes = @attributes[:user].with_indifferent_access. sacrés symboles ruby

Je suppose que tu parles du ` app/form_models/user_rdv_wizard.rb ` ?

**Comment by adipasquale** [comment]
> Je suppose que tu parles du `app/form_models/user_rdv_wizard.rb` ?

oui

**Comment by yaf** [comment]
Étonnant avec ce patch, si
- je crée un usager sans tel alors que c'est nécessaire
- je tente de continuer
- j'ai le message d'erreur comme quoi il manque le tel
- je pars modifier l'usager, je lui ajoute un numéro de téléphone
- quand je valide l'usager, je passe au step suivant (step3) avec une URL `rdv_wizard_step/new`

Je vais regarder ce qu'il se passe. Si tu as une idée, n'hésite pas à me le faire savoir :)

**Comment by adipasquale** [comment]
> avec ce patch

@yaf est-ce que tu fais reference au patch du `with_indifferent_access` ou au grand patch pour faire des formulaires GET au lieu de POST ?

**Comment by yaf** [comment]
Celui à propos de l'usage du GET au lieu du POST.
Je suis un peu surpris. Je n'ai pas encore compris pourquoi ça se comporte comme ça. Je cherche encore :)

**Comment by adipasquale** [comment]
bien vu @yaf j'avais raté ça. Ce qui se passe

- je crée un usager sans tel alors que c'est nécessaire
(route GET /new step2)

- je tente de continuer
-> on arrive sur la route GET /create step2

- j'ai le message d'erreur comme quoi il manque le tel
(toujours route la route GET /create step2)

- je pars modifier l'usager, je lui ajoute un numéro de téléphone
(toujours la route GET /create step2)

- quand je valide l'usager
-> à ce moment la modal refresh la page avec l'url du request.referer, c'est à dire la route GET /create step2, donc ça fait passer à l'étape 3.

Je n'ai pas trop de bonne solution pour ce sujet, j'imagine qu'on est obligés de passer explicitement l'URL on voudrait que ça retourne à la modale, plutôt que vers le request.referer. Ou bien de faire un redirect comme tu faisais avec le error_path , mais je suis moins fan de ça car ça fait un peu perdre le contexte

Du coup désolé ma suggestion ne semble pas résoudre le problème initial. On peut pousser un peu dans ce sens et je peux t'aider avec plaisir mais on peut aussi décaler ça à plus tard et partir avec ta solution du return_path

**Comment by yaf** [comment]
Ça me va de creuser un peu.

J'aime bien l'idée du GET, car finalement, ce que nous faisons, c'est enrichir un objet non persisté sur le chemin. D'ailleurs, ça pourrait être un update plutôt qu'un create. Est-ce que ça changerais un truc dans le cheminement :thinking:

**Comment by adipasquale** [comment]
pour ref 

```
diff --git a/app/controllers/admin/rdv_wizard_steps_controller.rb b/app/controllers/admin/rdv_wizard_steps_controller.rb
index 5cc5107c..6f7287ee 100644
--- a/app/controllers/admin/rdv_wizard_steps_controller.rb
+++ b/app/controllers/admin/rdv_wizard_steps_controller.rb
@@ -22,8 +22,6 @@ class Admin::RdvWizardStepsController < AgentAuthController
     authorize(@rdv_wizard.rdv, :create?)
     if @rdv_wizard.save
       redirect_to @rdv_wizard.success_path, @rdv_wizard.success_flash
-    elsif @rdv_wizard.error_path.present?
-      redirect_to @rdv_wizard.error_path, flash: { error: @rdv_wizard.errors.full_messages.join(", ") }
     else
       render current_step
     end
diff --git a/app/controllers/admin/users_controller.rb b/app/controllers/admin/users_controller.rb
index f1fe570f..bbc0127e 100644
--- a/app/controllers/admin/users_controller.rb
+++ b/app/controllers/admin/users_controller.rb
@@ -49,7 +49,8 @@ class Admin::UsersController < AgentAuthController
     prepare_new unless user_persisted
 
     if from_modal?
-      respond_modal_with @user, location: add_query_string_params_to_url(request.referer, 'user_ids[]': @user.id)
+      return_path = params[:return_path] || request.referer
+      respond_modal_with @user, location: add_query_string_params_to_url(return_path, 'user_ids[]': @user.id)
     elsif user_persisted
       flash[:notice] = "L'usager a été créé."
       redirect_to admin_organisation_user_path(@organisation, @user)
@@ -66,6 +67,7 @@ class Admin::UsersController < AgentAuthController
 
   def edit
     authorize(@user)
+    @return_path = params[:return_path]
     respond_modal_with @user if from_modal?
   end
 
@@ -74,7 +76,8 @@ class Admin::UsersController < AgentAuthController
     @user.skip_reconfirmation! if @user.encrypted_password.blank?
     user_updated = @user.update(user_params)
     flash[:notice] = "L'usager a été modifié." if user_updated
-    return respond_modal_with @user, location: request.referer if from_modal?
+    return_path = params[:return_path] || request.referer
+    return respond_modal_with @user, location: return_path if from_modal?
 
     if user_updated
       redirect_to admin_organisation_user_path(current_organisation, @user)
diff --git a/app/form_models/admin/rdv_wizard_form.rb b/app/form_models/admin/rdv_wizard_form.rb
index 960dd82b..fa8c7886 100644
--- a/app/form_models/admin/rdv_wizard_form.rb
+++ b/app/form_models/admin/rdv_wizard_form.rb
@@ -69,10 +69,6 @@ module Admin::RdvWizardForm
     def success_path
       new_admin_organisation_rdv_wizard_step_path(@organisation, step: 3, **to_query)
     end
-
-    def error_path
-      new_admin_organisation_rdv_wizard_step_path(@organisation, step: 2, **to_query)
-    end
   end
 
   class Step3 < Step2
diff --git a/app/views/admin/rdv_wizard_steps/step1.html.slim b/app/views/admin/rdv_wizard_steps/step1.html.slim
index 82e8c4e4..6286d0fb 100644
--- a/app/views/admin/rdv_wizard_steps/step1.html.slim
+++ b/app/views/admin/rdv_wizard_steps/step1.html.slim
@@ -14,7 +14,8 @@
               = "Commence le : #{l(@rdv.starts_at, format: :human)}"
       .card-body
         = render partial: 'layouts/model_errors', locals: { model: @rdv_wizard }
-        = simple_form_for(@rdv_wizard, as: :rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation, step: 1)) do |f|
+        = simple_form_for(@rdv_wizard, as: :rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation), method: :get) do |f|
+          = hidden_field_tag "step", "1"
           = f.input_field :starts_at, as: :hidden
           = f.input_field :lieu_id, as: :hidden
           = f.input_field :context, as: :hidden
diff --git a/app/views/admin/rdv_wizard_steps/step2.html.slim b/app/views/admin/rdv_wizard_steps/step2.html.slim
index 5fb7b91a..79c0fb0c 100644
--- a/app/views/admin/rdv_wizard_steps/step2.html.slim
+++ b/app/views/admin/rdv_wizard_steps/step2.html.slim
@@ -23,7 +23,8 @@
           label Usagers
           .js-users-spinner.spinner-border.d-none &nbsp;
 
-        = simple_form_for(@rdv_wizard, as: :rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation, step: 2)) do |f|
+        = simple_form_for(@rdv_wizard, as: :rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation), method: :get) do |f|
+          = hidden_field_tag "step", "2"
           - @rdv.users.each do |user|
             / https://stackoverflow.com/questions/2327621/creating-a-has-many-association-in-a-hidden-field
             = f.hidden_field "user_ids][", value: user.id
@@ -36,7 +37,7 @@
                   = relative_tag(user)
                 div.ml-2= link_to "╳ Retirer", "#", class: "js-remove-user", data: { 'user-id': user.id }
                 div.ml-2
-                  = link_to "✎ Modifier usager", edit_admin_organisation_user_path(current_organisation, user, modal: true), data: { modal: "true" }
+                  = link_to "✎ Modifier usager", edit_admin_organisation_user_path(current_organisation, user, modal: true, return_path: new_admin_organisation_rdv_wizard_step_path(@rdv_wizard.to_query.merge(step: 2))), data: { modal: "true" }
 
           .pl-2.mt-2= link_to "+ Ajouter un autre usager", "#", class: "js-toggle-add-user-interface #{'d-none' if @rdv.users.empty?}"
 
diff --git a/app/views/admin/rdv_wizard_steps/step3.html.slim b/app/views/admin/rdv_wizard_steps/step3.html.slim
index 6212a288..b3000167 100644
--- a/app/views/admin/rdv_wizard_steps/step3.html.slim
+++ b/app/views/admin/rdv_wizard_steps/step3.html.slim
@@ -17,8 +17,9 @@
           = users_inline_list_for_agents(@rdv.users)
 
       .card-body
-        = simple_form_for(@rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation, step: 3)) do |f|
+        = simple_form_for(@rdv, url: admin_organisation_rdv_wizard_step_path(current_organisation), method: :get) do |f|
           = render(partial: 'layouts/model_errors', locals: { model: @rdv_wizard, f: f })
+          = hidden_field_tag "step", "3"
           = collapsible_form_fields_for_warnings(@rdv_wizard) do
             = f.association :motif, as: :hidden,  wrapper_html: { class: 'm-0' }
             = f.input :context, as: :hidden, wrapper_html: { class: 'm-0' }
diff --git a/app/views/admin/users/_form.html.slim b/app/views/admin/users/_form.html.slim
index 811a951d..0783deca 100644
--- a/app/views/admin/users/_form.html.slim
+++ b/app/views/admin/users/_form.html.slim
@@ -6,7 +6,6 @@
   user, url: agent_user_form_url(user), remote: from_modal?, \
   html: { class: "js-agent-user-form" }, data: { modal: from_modal? } \
 ) do |f|
-
   - f.simple_fields_for :user_profiles, user.profile_for(current_organisation) do |f_user_profile|
     = f_user_profile.input :organisation_id, as: :hidden, wrapper: false
 
@@ -110,6 +109,9 @@
       - else
         = link_to "Annuler", agent_user_form_cancel_path(user), class: "btn btn-link"
 
+      - if @return_path.present?
+        = hidden_field_tag :return_path, @return_path
+
       - if user.persisted?
         = f.button :submit
       - else
diff --git a/config/routes.rb b/config/routes.rb
index d65e048e..a8888ca7 100644
--- a/config/routes.rb
+++ b/config/routes.rb
@@ -143,7 +143,9 @@ Rails.application.routes.draw do
           post :reinvite, on: :member
         end
         resource :merge_users, only: [:new, :create]
-        resource :rdv_wizard_step, only: [:new, :create]
+        resource :rdv_wizard_step, only: [:new] do
+          get :create
+        end
         devise_for :agents, controllers: { invitations: "admin/invitations_devise" }, only: :invitations
       end
 

```

**Comment by yaf** [comment]
Comme évoquer durant notre session de binôme, je reste sur le schéma que j'avais commencé. Je pense qu'a long terme, nous pourrions essayer de supprimer complétement ce système de « modal » qui est un peu bizarre...

**Review by adipasquale — CHANGES_REQUESTED** [review]
déso je note encore un souci : coté agent si tu tentes d'avancer sans avoir choisi d'usager il y a une 500 parce que tu fais `users.first.phone_number` => users est vide a ce moment. tu peux utiliser le safe operator pour ce cas &.

File: app/views/users/users/_form.html.slim [hunk 1]
```diff
@@ -8,12 +8,12 @@
...
     .col-md-6= f.input :birth_name, placeholder: 'Nom de naissance'
     .col-md-6= date_input(f, :birth_date)
-  = f.input :phone_number, as: :tel
+  = f.input :phone_number, as: :tel, required: local_assigns[:rdv_wizard]&.motif&.phone?
   label Préférences de notifications
   .d-flex
     div= f.input :notify_by_email
     div.ml-3= f.input :notify_by_sms
-  - address_value = local_assigns[:rdv_wizard].present? && user.address.nil? ? rdv_wizard[:where] : user.address
+  - address_value = local_assigns[:rdv_wizard].present? && user.address.nil? ? rdv_wizard.to_query[:where] : user.address
```
Comment (adipasquale) [inline app/views/users/users/_form.html.slim:16]: en passant je note qu'on pourrait eviter de passer par to_query ici meme si ce n'est pas toi qui l'a introduit

Comment (yaf) [reply app/views/users/users/_form.html.slim:16]: J'en profite pour corriger, merci !

**Comment by yaf** [comment]
Je crois que c'est bon. Il y a eu un pépin sur un test lié à la recherche (flaky test ?)

File: app/views/users/users/_form.html.slim (hunk 1 above)
Comment (yaf) [reply app/views/users/users/_form.html.slim:16]: Finalement non, c'est pas possible de l'enlever. Je le remets. C'est ça qui fait exploser le test actuellement.

Comment (adipasquale) [reply app/views/users/users/_form.html.slim:16]: oui je voulais dire qu'il faudrait rajouter un `delegate :where, to :attributes` ou qqch comme ca dans le modèle pour pouvoir faire `rdv_wizard.where` plutot que passer par la serialisation en to_query puis l'acces a la clé where

C'est un detail donc pas de soucis pour faire ca plus tard

**Review by adipasquale — CHANGES_REQUESTED** [review]
bon désolé yannick je ne fais pas du tout expres que ca soit aussi iteratif, mais il y a encore un souci 

en fait ton redirect en cas d'erreur semble aussi se produire sur l'etape 3 par ex et ca casse le formatting des warnings : 

Screenshot 2021-02-15 at 17 53 59

un peu la tete dans le guidon en fin de journee la mais je n'ai pas trop d'autres pistes que le contournement propsoé avec les routes GET pour supprimer le redirect_to error_path. on peut en rediscuter demain

**Comment by yaf** [comment]
Arg. C'est sans doute lié à l'héritage empilé :up: 

Il faudrait repartir sur le `GET` et l'usage du `return_path`. Bon, je vais essayer de faire ça. Merci d'avoir testé correctement !

**Comment by adipasquale** [comment]
comme discuté hier @yaf je me suis permis de commiter directement plutot que t'envoyer un patch mais n'hesite pas a le retoucher (avec des commits successifs plutot que reecrire l'histoirque, ca complique un peu la collaboration sinon)

- il fallait faire la meme manip pour passer le return location dans le cas d'une creation d'user depuis la modale
- j'ai explosé les lignes des appels aux modales qui devenaient durs à lire dans la vue
- j'ai corrigé un edge case si c'est pas le premier user mais un suivant qui a un num de tél

**Comment by yaf** [comment]
Merci.

J'ai l'impression que c'est le « moins pire » que nous pouvons faire :-/

**Review by adipasquale — APPROVED** [review]
looks good!
